catt/controllers.py

```diff
@@ -92,50 +63,81 @@ def setup_cast(device_name, video_url=None, prep=None, controller=None, ytdl_opt
     except (pychromecast.error.ChromecastConnectionError, ValueError):
         cast = get_chromecast(device_name)
         cache.set_data(cast.name, cast.host, cast.port)
+
     cast.wait()
+    return cast
 
-    if video_url:
-        model_name = DEVICES_WITH_TWO_MODEL_NAMES.get(cast.model_name, cast.model_name)
-        cc_info = (cast.device.manufacturer, model_name)
-        stream = StreamInfo(
-            video_url, model=cc_info, host=cast.host, device_type=cast.cast_type, ytdl_options=ytdl_options
-        )
 
-    if controller:
-        if controller == "default":
-            app = DEFAULT_APP
-        else:
-            app = next(a for a in APP_INFO if a["app_name"] == controller)
-    elif stream and prep == "app":
-        if stream.is_local_file:
-            app = DEFAULT_APP
+def get_stream(url, device_info=None, host=None, ytdl_options=None):
+    if device_info:
+        model_name = DEVICES_WITH_TWO_MODEL_NAMES.get(device_info.model_name, device_info.model_name)
+        cc_info = (device_info.manufacturer, model_name)
+        cast_type = device_info.cast_type
+    else:
+        cc_info = cast_type = None
+    return StreamInfo(url, host=host, model=cc_info, device_type=cast_type, ytdl_options=ytdl_options)
+
+
+def get_app_info(id_or_name, cast_type=None, strict=False, show_warning=False):
+    if id_or_name == "default":
+        return DEFAULT_APP
+
+    field = "app_id" if re.match("[0-9A-F]{8}$", id_or_name) else "app_name"
+    try:
+        app_info = next(a for a in APP_INFO if a[field] == id_or_name)
+    except StopIteration:
+        if strict:
+            raise AppSelectionError("app not found (strict is set)")
         else:
-            try:
-                app = next(a for a in APP_INFO if a["app_name"] == stream.extractor)
-            except StopIteration:
-                app = DEFAULT_APP
+            app_info = DEFAULT_APP
+
+    if app_info["app_name"] != "default":
+        if not cast_type:
+            raise AppSelectionError("cast_type is needed for app selection")
+        elif cast_type not in app_info["supported_device_types"]:
+            msg = "The %s app is not available for this device." % app_info["app_name"].capitalize()
+            if strict:
+                raise CattCastError(msg)
+            elif show_warning:
+                warning(msg)
+            app_info = DEFAULT_APP
+    return app_info
+
+
+def get_controller(cast, app_info, action=None, prep=None):
+    app_name = app_info["app_name"]
```

Comment:
This is a bit more obfuscated, but it should be clear enough, and the code becomes cleaner with multiple controllers:

```suggestion
    controller = {
        "youtube": YoutubeCastController,
        "dashcast": DashCastController,
    }.get(app_info["app_name"], DashCastController)
```

Reply:
Cheers